File: classified/packages/plugin-stagehand/src/websocket-client.ts
import { logger } from '@elizaos/core';

// Dynamic import to handle both Node.js and bundled environments
let WebSocket: typeof import('ws').default;
if (typeof window !== 'undefined' && typeof window.WebSocket !== 'undefined') {
  // Browser environment
  WebSocket = window.WebSocket as any;
} else {
  // Node.js environment
  WebSocket = require('ws');
}

export interface StagehandMessage {
  type: string;
  requestId?: string;
  [key: string]: any;
}

export class StagehandWebSocketClient {
  private ws: any = null; // Using any to avoid TypeScript issues with ws module
  private messageHandlers = new Map<string, (response: StagehandMessage) => void>();
  private connected = false;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;

  constructor(private serverUrl: string) {}

  async connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        this.ws = new (WebSocket as any)(this.serverUrl);

        this.ws.on('open', () => {
          this.connected = true;
          this.reconnectAttempts = 0;
          logger.info(`[Stagehand] Connected to server at ${this.serverUrl}`);
          resolve();
        });

        this.ws.on('message', (data: any) => {
          try {
            const message = JSON.parse(data.toString()) as StagehandMessage;

            // Handle response messages with requestId
            if (message.requestId && this.messageHandlers.has(message.requestId)) {
              const handler = this.messageHandlers.get(message.requestId)!;
              this.messageHandlers.delete(message.requestId);
              handler(message);
            }

            // Log other messages
            if (message.type === 'connected') {
              logger.info(`[Stagehand] Server connected: ${JSON.stringify(message)}`);
            }
          } catch (error) {
            logger.error('[Stagehand] Error parsing message:', error);
          }
        });

        this.ws.on('error', (error: any) => {
          logger.error('[Stagehand] WebSocket error:', error);
          if (!this.connected) {
            reject(error);
          }
        });

        this.ws.on('close', () => {
          this.connected = false;
          logger.info('[Stagehand] Disconnected from server');

          // Attempt reconnection if not explicitly disconnected
          if (this.ws && this.reconnectAttempts < this.maxReconnectAttempts) {
            this.attemptReconnect();
          }
        });
      } catch (error) {
        reject(error);
      }
    });
  }

  private async attemptReconnect(): Promise<void> {
    this.reconnectAttempts++;
    logger.info(
      `[Stagehand] Attempting reconnection ${this.reconnectAttempts}/${this.maxReconnectAttempts}...`
    );

    await new Promise((resolve) =>
      setTimeout(resolve, this.reconnectDelay * this.reconnectAttempts)
    );

    try {
      await this.connect();
    } catch (error) {
      logger.error('[Stagehand] Reconnection failed:', error);
    }
  }

  async sendMessage(type: string, data: any): Promise<StagehandMessage> {
    if (!this.ws || !this.connected) {
      throw new Error('Not connected to Stagehand server');
    }

    const requestId = `req-${Date.now()}-${Math.random().toString(36).substring(7)}`;
    const message: StagehandMessage = {
      type,
      requestId,
      ...data,
    };

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.messageHandlers.delete(requestId);
        reject(new Error(`Request timeout for ${type}`));
      }, 30000); // 30 second timeout

      this.messageHandlers.set(requestId, (response) => {
        clearTimeout(timeout);
        if (response.type === 'error') {
          reject(new Error(response.error || 'Unknown error'));
        } else {
          resolve(response);
        }
      });

      this.ws!.send(JSON.stringify(message));
      logger.debug(`[Stagehand] Sent message: ${type} (${requestId})`);
    });
  }

  disconnect(): void {
    this.reconnectAttempts = this.maxReconnectAttempts; // Prevent reconnection
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.connected = false;
    logger.info('[Stagehand] Client disconnected');
  }

  isConnected(): boolean {
    return this.connected;
  }

  // Convenience methods for specific actions

  async navigate(sessionId: string, url: string): Promise<{ url: string; title: string }> {
    const response = await this.sendMessage('navigate', {
      sessionId,
      data: { url },
    });
    return response.data || { url, title: '' };
  }

  async getState(sessionId: string): Promise<{
    url: string;
    title: string;
    sessionId: string;
    createdAt: Date;
  }> {
    const response = await this.sendMessage('getState', { sessionId });
    return (
      response.data || {
        url: '',
        title: '',
        sessionId,
        createdAt: new Date(),
      }
    );
  }

  async goBack(sessionId: string): Promise<{ url: string; title: string }> {
    const response = await this.sendMessage('goBack', { sessionId });
    return response.data || { url: '', title: '' };
  }

  async goForward(sessionId: string): Promise<{ url: string; title: string }> {
    const response = await this.sendMessage('goForward', { sessionId });
    return response.data || { url: '', title: '' };
  }

  async refresh(sessionId: string): Promise<{ url: string; title: string }> {
    const response = await this.sendMessage('refresh', { sessionId });
    return response.data || { url: '', title: '' };
  }

  async click(sessionId: string, description: string): Promise<StagehandMessage> {
    return await this.sendMessage('click', {
      sessionId,
      data: { description },
    });
  }

  async type(sessionId: string, text: string, field: string): Promise<StagehandMessage> {
    return await this.sendMessage('type', {
      sessionId,
      data: { text, field },
    });
  }

  async select(sessionId: string, option: string, dropdown: string): Promise<StagehandMessage> {
    return await this.sendMessage('select', {
      sessionId,
      data: { option, dropdown },
    });
  }

  async extract(sessionId: string, instruction: string): Promise<StagehandMessage> {
    return await this.sendMessage('extract', {
      sessionId,
      data: { instruction },
    });
  }

  async screenshot(sessionId: string): Promise<StagehandMessage> {
    return await this.sendMessage('screenshot', { sessionId });
  }

  async solveCaptcha(sessionId: string): Promise<StagehandMessage> {
    return await this.sendMessage('solveCaptcha', { sessionId });
  }

  async health(): Promise<boolean> {
    try {
      const response = await this.sendMessage('health', {});
      return response.type === 'health' && response.data?.status === 'ok';
    } catch (error) {
      logger.error('[Stagehand] Health check failed:', error);
      return false;
    }
  }
}

File: cpp_merge/classified/packages/plugin-stagehand/src/websocket-client.cpp
#include "websocket-client.h"
#include <string>

StagehandWebSocketClient::StagehandWebSocketClient(string serverUrl_) : serverUrl(serverUrl_)  {
}

std::shared_ptr<Promise<void>> StagehandWebSocketClient::connect()
{
    return std::make_shared<Promise>([=](auto resolve, auto reject) mutable
    {
        try
        {
            this->ws = std::make_shared<(as<any>(WebSocket))>(this->serverUrl);
            this->ws["on"](std::string("open"), [=]() mutable
            {
                this->connected = true;
                this->reconnectAttempts = 0;
                logger->info(std::string("[Stagehand] Connected to server at ") + this->serverUrl + string_empty);
                resolve();
            }
            );
            this->ws["on"](std::string("message"), [=](auto data) mutable
            {
                try
                {
                    auto message = as<std::shared_ptr<StagehandMessage>>(JSON->parse(data["toString"]()));
                    if (AND((message->requestId), (this->messageHandlers->has(message->requestId)))) {
                        auto handler = this->messageHandlers->get(message->requestId);
                        this->messageHandlers->delete(message->requestId);
                        handler(message);
                    }
                    if (message->type == std::string("connected")) {
                        logger->info(std::string("[Stagehand] Server connected: ") + JSON->stringify(message) + string_empty);
                    }
                }
                catch (const any& error)
                {
                    logger->error(std::string("[Stagehand] Error parsing message:"), error);
                }
            }
            );
            this->ws["on"](std::string("error"), [=](auto error) mutable
            {
                logger->error(std::string("[Stagehand] WebSocket error:"), error);
                if (!this->connected) {
                    reject(error);
                }
            }
            );
            this->ws["on"](std::string("close"), [=]() mutable
            {
                this->connected = false;
                logger->info(std::string("[Stagehand] Disconnected from server"));
                if (AND((this->ws), (this->reconnectAttempts < this->maxReconnectAttempts))) {
                    this->attemptReconnect();
                }
            }
            );
        }
        catch (const any& error)
        {
            reject(error);
        }
    }
    );
}

std::shared_ptr<Promise<void>> StagehandWebSocketClient::attemptReconnect()
{
    this->reconnectAttempts++;
    logger->info(std::string("[Stagehand] Attempting reconnection ") + this->reconnectAttempts + std::string("/") + this->maxReconnectAttempts + std::string("..."));
    std::async([=]() { std::make_shared<Promise>([=](auto resolve) mutable
    {
        return setTimeout(resolve, this->reconnectDelay * this->reconnectAttempts);
    }
    ); });
    try
    {
        std::async([=]() { this->connect(); });
    }
    catch (const any& error)
    {
        logger->error(std::string("[Stagehand] Reconnection failed:"), error);
    }
    return std::shared_ptr<Promise<void>>();
}

std::shared_ptr<Promise<std::shared_ptr<StagehandMessage>>> StagehandWebSocketClient::sendMessage(string type, any data)
{
    if (OR((!this->ws), (!this->connected))) {
        throw any(std::make_shared<Error>(std::string("Not connected to Stagehand server")));
    }
    shared requestId = std::string("req-") + Date->now() + std::string("-") + Math->random()->toString(36)->substring(7) + string_empty;
    shared message = utils::assign(object{
        object::pair{std::string("type"), std::string("type")}, 
        object::pair{std::string("requestId"), std::string("requestId")}
    }, data);
    return std::make_shared<Promise>([=](auto resolve, auto reject) mutable
    {
        shared timeout = setTimeout([=]() mutable
        {
            this->messageHandlers->delete(requestId);
            reject(std::make_shared<Error>(std::string("Request timeout for ") + type + string_empty));
        }
        , 30000);
        this->messageHandlers->set(requestId, [=](auto response) mutable
        {
            clearTimeout(timeout);
            if (response->type == std::string("error")) {
                reject(std::make_shared<Error>(OR((response->error), (std::string("Unknown error")))));
            } else {
                resolve(response);
            }
        }
        );
        this->ws["send"](JSON->stringify(message));
        logger->debug(std::string("[Stagehand] Sent message: ") + type + std::string(" (") + requestId + std::string(")"));
    }
    );
}

void StagehandWebSocketClient::disconnect()
{
    this->reconnectAttempts = this->maxReconnectAttempts;
    if (this->ws) {
        this->ws["close"]();
        this->ws = nullptr;
    }
    this->connected = false;
    logger->info(std::string("[Stagehand] Client disconnected"));
}

boolean StagehandWebSocketClient::isConnected()
{
    return this->connected;
}

std::shared_ptr<Promise<object>> StagehandWebSocketClient::navigate(string sessionId, string url)
{
    auto response = std::async([=]() { this->sendMessage(std::string("navigate"), object{
        object::pair{std::string("sessionId"), std::string("sessionId")}, 
        object::pair{std::string("data"), object{
            object::pair{std::string("url"), std::string("url")}
        }}
    }); });
    return OR((response->data), (object{
        object::pair{std::string("url"), std::string("url")}, 
        object::pair{std::string("title"), string_empty}
    }));
}

std::shared_ptr<Promise<object>> StagehandWebSocketClient::getState(string sessionId)
{
    auto response = std::async([=]() { this->sendMessage(std::string("getState"), object{
        object::pair{std::string("sessionId"), std::string("sessionId")}
    }); });
    return (OR((response->data), (object{
        object::pair{std::string("url"), string_empty}, 
        object::pair{std::string("title"), string_empty}, 
        object::pair{std::string("sessionId"), std::string("sessionId")}, 
        object::pair{std::string("createdAt"), std::make_shared<Date>()}
    })));
}

std::shared_ptr<Promise<object>> StagehandWebSocketClient::goBack(string sessionId)
{
    auto response = std::async([=]() { this->sendMessage(std::string("goBack"), object{
        object::pair{std::string("sessionId"), std::string("sessionId")}
    }); });
    return OR((response->data), (object{
        object::pair{std::string("url"), string_empty}, 
        object::pair{std::string("title"), string_empty}
    }));
}

std::shared_ptr<Promise<object>> StagehandWebSocketClient::goForward(string sessionId)
{
    auto response = std::async([=]() { this->sendMessage(std::string("goForward"), object{
        object::pair{std::string("sessionId"), std::string("sessionId")}
    }); });
    return OR((response->data), (object{
        object::pair{std::string("url"), string_empty}, 
        object::pair{std::string("title"), string_empty}
    }));
}

std::shared_ptr<Promise<object>> StagehandWebSocketClient::refresh(string sessionId)
{
    auto response = std::async([=]() { this->sendMessage(std::string("refresh"), object{
        object::pair{std::string("sessionId"), std::string("sessionId")}
    }); });
    return OR((response->data), (object{
        object::pair{std::string("url"), string_empty}, 
        object::pair{std::string("title"), string_empty}
    }));
}

std::shared_ptr<Promise<std::shared_ptr<StagehandMessage>>> StagehandWebSocketClient::click(string sessionId, string description)
{
    return std::async([=]() { this->sendMessage(std::string("click"), object{
        object::pair{std::string("sessionId"), std::string("sessionId")}, 
        object::pair{std::string("data"), object{
            object::pair{std::string("description"), std::string("description")}
        }}
    }); });
}

std::shared_ptr<Promise<std::shared_ptr<StagehandMessage>>> StagehandWebSocketClient::type(string sessionId, string text, string field)
{
    return std::async([=]() { this->sendMessage(std::string("type"), object{
        object::pair{std::string("sessionId"), std::string("sessionId")}, 
        object::pair{std::string("data"), object{
            object::pair{std::string("text"), std::string("text")}, 
            object::pair{std::string("field"), std::string("field")}
        }}
    }); });
}

std::shared_ptr<Promise<std::shared_ptr<StagehandMessage>>> StagehandWebSocketClient::select(string sessionId, string option, string dropdown)
{
    return std::async([=]() { this->sendMessage(std::string("select"), object{
        object::pair{std::string("sessionId"), std::string("sessionId")}, 
        object::pair{std::string("data"), object{
            object::pair{std::string("option"), std::string("option")}, 
            object::pair{std::string("dropdown"), std::string("dropdown")}
        }}
    }); });
}

std::shared_ptr<Promise<std::shared_ptr<StagehandMessage>>> StagehandWebSocketClient::extract(string sessionId, string instruction)
{
    return std::async([=]() { this->sendMessage(std::string("extract"), object{
        object::pair{std::string("sessionId"), std::string("sessionId")}, 
        object::pair{std::string("data"), object{
            object::pair{std::string("instruction"), std::string("instruction")}
        }}
    }); });
}

std::shared_ptr<Promise<std::shared_ptr<StagehandMessage>>> StagehandWebSocketClient::screenshot(string sessionId)
{
    return std::async([=]() { this->sendMessage(std::string("screenshot"), object{
        object::pair{std::string("sessionId"), std::string("sessionId")}
    }); });
}

std::shared_ptr<Promise<std::shared_ptr<StagehandMessage>>> StagehandWebSocketClient::solveCaptcha(string sessionId)
{
    return std::async([=]() { this->sendMessage(std::string("solveCaptcha"), object{
        object::pair{std::string("sessionId"), std::string("sessionId")}
    }); });
}

std::shared_ptr<Promise<boolean>> StagehandWebSocketClient::health()
{
    try
    {
        auto response = std::async([=]() { this->sendMessage(std::string("health"), object{}); });
        return AND((response->type == std::string("health")), (response->data["status"] == std::string("ok")));
    }
    catch (const any& error)
    {
        logger->error(std::string("[Stagehand] Health check failed:"), error);
        return false;
    }
}

any WebSocket;

void Main(void)
{
    if (AND((type_of(window) != std::string("std::nullopt")), (type_of(window->WebSocket) != std::string("std::nullopt")))) {
        WebSocket = as<any>(window->WebSocket);
    } else {
        WebSocket = require(std::string("ws"));
    }
}

MAIN
